astropy/cosmology/core.py

```diff
@@ -166,23 +354,102 @@ def __eq__(self, other):
         Returns
         -------
         bool
-            True if all immutable fields are equal, False otherwise.
-
+            True if cosmologies are equivalent, False otherwise.
         """
+        # The options are: 1) same class & parameters; 2) same class, different
+        # parameters; 3) different classes, equivalent parameters; 4) different
+        # classes, different parameters. (1) & (3) => True, (2) & (4) => False.
+
         if not isinstance(other, Cosmology):
             return False
+        # passed this basic check, so 'other' should have method '__equiv__'
 
-        sias, oias = self._init_arguments, other._init_arguments
+        equiv = self.__equiv__(other)  # first try from 'self'
 
-        # check if the cosmologies have identical signatures.
-        # this protects against one cosmology having a superset of input
-        # parameters to another cosmology.
-        if (sias.keys() ^ oias.keys()) - {'meta'}:
-            return False
+        if equiv is NotImplemented:  # that failed, try starting from 'other'
+            equiv = other.__equiv__(self)
+
+        return equiv if equiv is not NotImplemented else False
+
+    def __equiv__(self, other):
+        """Cosmology equivalence. Use ``.is_equivalent()`` for actual check!
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.Cosmology` subclass instance
+            The object in which to compare.
 
-        # are all the non-excluded immutable arguments equal?
-        return all((np.all(oias[k] == v) for k, v in sias.items()
-                    if k != "meta"))
+        Returns
+        -------
+        bool or `NotImplemented`
+            `NotImplemented` if 'other' is from a different class.
+            `True` if 'other' is of the same class and has matching parameters
+            and parameter values. `False` otherwise.
+        """
+        if other.__class__ is not self.__class__:
+            return NotImplemented  # allows other.__equiv__
+
+        # check all parameters in other match those in 'self' and 'other' has
+        # no extra parameters (latter part should never happen b/c same class)
+        params_eq = (set(self.__parameters__) == set(other.__parameters__)
+                     and all(np.all(getattr(self, k) == getattr(other, k))
+                             for k in self.__parameters__))
+        return params_eq
+
+    def __eq__(self, other):
+        """Check equality between Cosmologies.
+
+        Checks the Parameters and immutable fields (i.e. not "meta").
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.Cosmology` subclass instance
+            The object in which to compare.
+
+        Returns
+        -------
+        bool
+            True if Parameters and names are the same, False otherwise.
+        """
+        if other.__class__ is not self.__class__:
```

Comment:
Hmm, since this first part is identical to `__equiv__`, maybe just 
```
param_eq = Cosmology.__equiv__(self, other)
```
(Here, one does indeed have to use `Cosmolgy` since a subclass may decide differently)

Reply:
Haha. I had implemented it this way before and have been oscillating back and forth about whether it was cleaner or not to repeat the code. Your vote tilts it to `Cosmology.__equiv__`

Reply:
Oh wait. I forgot it's not quite so simple. Because ``NotImplemented`` should be returned, it actually needs to be

```
param_eq = Cosmology.__equiv__(self, other)
if param_eq is NotImplemented:
    return NotImplemented
```

Which is not much simpler than the current implementation.

Reply:
Dunno. The below still feels a bit cleaner:
```
if equivalent := Cosmology.__equiv__(self, other) is NotImplemented:
    return NotImplemented
else:
    return equivalent and self.name == other.name
```
(hey, I got a walrus in!)

Though, actually in terms of actual logic, maybe it is best to state that by default equality equals same class, same name, and equivalent? And rely on equivalency always being defined sensibly, i.e.,
```
if self.__class is not other.__class__:
    return NotImplemented

return self.name == other.name and self.__equiv__(other)
```

Reply:
I like the clarity of the 2nd version. The class check is inexpensive, so it's fine to repeat.
I'm keeping the variable assignment because I'm thinking off adding a strict equality that includes metadata. Cosmology aren't singletons but it would be nice to have a method that mirrors ``x is y`` for checking absolute equality.

```
# check all parameters in 'other' match those in 'self'
equivalent = self.__equiv__(other)
# non-Parameter checks: name
name_eq = (self.name == other.name)
```